ntb_hw_switchtec.c

```diff
@@ -1573,6 +1617,146 @@ static int switchtec_ntb_reinit_peer(struct switchtec_ntb *sndev)
+	ssize_t (*show)(struct switchtec_ntb *, char *);
+	ssize_t (*store)(struct switchtec_ntb *, const char *, size_t);
+};
+
+static ssize_t requester_ids_show(struct switchtec_ntb *sndev, char *page)
+{
+	int i;
+	int req_id_table_size;
+	char req_id_str[32];
+	u32 req_id;
+	ssize_t n = 0;
+
+	req_id_table_size = ioread16(&sndev->mmio_self_ctrl->req_id_table_size);
+
+	for (i = 0; i < req_id_table_size; i++) {
+		req_id = ioread32(&sndev->mmio_self_ctrl->req_id_table[i]);
+
+		if (!(req_id & NTB_CTRL_REQ_ID_EN))
+			break;
+
+		req_id >>= 16;
+		n += sprintf(req_id_str, "Requester ID #%d: %02x:%02x.%x\n", i,
+			     PCI_BUS_NUM(req_id), PCI_SLOT(req_id),
+			     PCI_FUNC(req_id));
+		strcat(page, req_id_str);
+	}
+
+	return n;
+}
+
+static ssize_t requester_ids_store(struct switchtec_ntb *sndev,
+				   const char *page, size_t count)
+{
+	int bus, dev, func;
+	int req_id;
+	bool add_req_id;
+	ssize_t ret;
+
+	if (sscanf(page, "add %x:%x.%x", &bus, &dev, &func) == 3) {
+		add_req_id = true;
+	} else if (sscanf(page, "del %x:%x.%x", &bus, &dev, &func) == 3) {
+		add_req_id = false;
+	} else {
+		return -EINVAL;
+	}
+
+	if ((bus > 0xff) || (dev > 0x1f) || (func > 8))
+		ret = -EINVAL;
+
+	req_id = PCI_DEVID(bus, PCI_DEVFN(dev, func));
+
+	if (add_req_id) {
+		ret = add_req_ids(sndev, sndev->mmio_self_ctrl, &req_id, 1);
+		if (ret)
+			return ret;
+	} else {
+		ret = del_req_id(sndev, sndev->mmio_self_ctrl, req_id);
+		if (ret)
+			return ret;
+	}
+
+	if (crosslink_is_enabled(sndev)) {
+		ret = crosslink_setup_req_ids(sndev, sndev->mmio_peer_ctrl);
+		if (ret)
+			return ret;
+	}
+
+	return count;
+}
+struct ntb_sysfs_entry requester_ids_attr = __ATTR_RW(requester_ids);
+
+static ssize_t switchtec_ntb_attr_show(struct kobject *kobj,
+				       struct attribute *attr, char *page)
+{
+	struct ntb_sysfs_entry *entry;
+	struct switchtec_ntb *sndev;
+
+	entry = container_of(attr, struct ntb_sysfs_entry, attr);
+	sndev = container_of(kobj, struct switchtec_ntb, kobj);
+
+	if (!entry->show)
+		return -EIO;
+	return entry->show(sndev, page);
+}
+
+static ssize_t switchtec_ntb_attr_store(struct kobject *kobj,
+					struct attribute *attr,
+					const char *page, size_t count)
+{
+	struct ntb_sysfs_entry *entry;
+	struct switchtec_ntb *sndev;
+
+	entry = container_of(attr, struct ntb_sysfs_entry, attr);
+	sndev = container_of(kobj, struct switchtec_ntb, kobj);
+
+	if (!entry->store)
+		return -EIO;
+	return entry->store(sndev, page, count);
+}
+
+static struct sysfs_ops switchtec_ntb_sysfs_ops = {
+	.show   = switchtec_ntb_attr_show,
+	.store  = switchtec_ntb_attr_store,
+};
+
+static struct attribute *switchtec_ntb_attrs[] = {
+	&requester_ids_attr.attr,
+	NULL,
+};
+
+static struct kobj_type switchtec_ntb_ktype = {
+	.sysfs_ops = &switchtec_ntb_sysfs_ops,
+	.default_attrs = switchtec_ntb_attrs,
+};
+
+static void switchtec_ntb_init_sysfs(struct switchtec_ntb *sndev)
+{
+	struct kobject *parent;
+	int err;
+
+	parent = &sndev->stdev->dev.kobj;
```

Comment:
I'd probably add these attributes to the NTB device (sndev->ntb->dev) and not the PCI device.

Reply:
Agree, will update.